network-config-analyzer/IstioSidecarYamlParser.py

```diff
@@ -0,0 +1,196 @@
+        :rtype: IstioSidecarRule
+        """
+        # currently only hosts is considered in the rule parsing, other fields are not supported
+        allowed_keys = {'port': [3, dict], 'bind': [3, str], 'captureMode': [3, str], 'hosts': [1, list]}
+        self.check_fields_validity(egress_rule, 'Istio Egress Listener', allowed_keys)
+
+        hosts = egress_rule.get('hosts')
+        if not hosts:
+            self.syntax_error('One or more service hosts to be exposed by the listener are required', egress_rule)
+        res_peers = PeerSet()
+        special_res_peers = PeerSet()
+        for host in hosts:
+            # Services in the specified namespace matching dnsName will be exposed.
+            namespace, dns_name = self._validate_and_partition_host_format(host)
+            host_peers, special_case_host = self._get_peers_from_host_namespace(namespace)
+            self._validate_dns_name_pattern(dns_name)
+            if host_peers:  # if there are services in the specified namespace
+                if '*' not in dns_name:  # * means all services in namespace, all already in host_peers
+                    host_peers &= self._get_peers_from_host_dns_name(dns_name)
+            if special_case_host:
+                special_res_peers |= host_peers
+            else:
+                res_peers |= host_peers
+
+        return IstioSidecarRule(res_peers, special_res_peers)
+
+    def _check_and_save_sidecar_if_top_priority(self, curr_sidecar):
+        """
+        check if current sidecar is top priority for its namespace or workloads.
+        if the sidecar is selector less, and is the first default sidecar for current namespace, save it
+        otherwise, save it for any peer if this is the first sidecar selecting it
+        :param IstioSidecar curr_sidecar: the sidecar parsed in self
+        a warning message will be printed if current sidecar is not the first for its relevant object,
+        indicating that this sidecar will be ignored in the sidecar's connections
+        """
+        if curr_sidecar.default_sidecar:
+            if self.namespace.prior_default_sidecar:  # this sidecar is not first one
+                self.warning(f'Namespace "{str(self.namespace)}" already has a Sidecar configuration '
+                             f'without any workloadSelector. '
+                             f'Connections in sidecar: "{curr_sidecar.full_name()}" will be ignored')
+                return
+            self.namespace.prior_default_sidecar = curr_sidecar
+            return
+
+        for peer in curr_sidecar.selected_peers:
+            if peer.prior_sidecar:  # this sidecar is not first one
+                self.warning(f'Peer "{peer.full_name()}" already has a Sidecar configuration selecting it. Sidecar: '
+                             f'"{curr_sidecar.full_name()}" will not be considered as connections for this workload')
+                return
+            peer.prior_sidecar = curr_sidecar
```

Comment:
What cases does it cover of "another sidecar selecting this workload/namespace"? 
Shouldn't we issue error, for example, if two different sidecar objects select the same workload?
From istio documentation: 
"Each namespace can have only one Sidecar configuration without any workloadSelector that specifies the default for all pods in that namespace. It is recommended to use the name default for the namespace-wide sidecar. The behavior of the system is undefined if more than one selector-less Sidecar configurations exist in a given namespace. The behavior of the system is undefined if two or more Sidecar configurations with a workloadSelector select the same workload instance."

Reply:
On Live cluster, the "first" sidecar is considered always.  (they don't raise any errors / warnings)
So, we decided to raise a warning and consider the first sidecar (behave the same)

Reply:
what is the "first" sidecar?

Reply:
I mean, for example, if you have two sidecars selecting the same workloads.
On live cluster, the sidecar that you apply first will be the only one considered for those workloads' connections. 
on our code, the first of them listed in our  `networkPolicyList` for a config will be considered ( a warning will be printed to tell that the other sidecar will not be considered) 
examples are in `bookinfo-sidecar-tests-scheme.yaml` :

1.  two-sidecars-with-same-workload-selector, query: connectivity-map-bookinfo-two-sidecars-with-same-workload-selector
2. (for default sidecars): 
        bookinfo-two-selector-less-sidecars, query : connectivity-map-bookinfo-ignoring-second-default-sidecar-in-same-namespace